### PR TITLE
[`from_pretrained`] Make from_pretrained fast again

### LINKED ISSUE #18505: Suppress `reset_parameters` of `torch.nn.Linear,Conv2d...` inside `no_init_weights`
### Feature request

`torch.nn.Linear,Conv2d...` will call `self.reset_parameters()` inside their `__init__`.
I'd like to make `reset_parameters` be a no-op inside `no_init_weights` context manager.

### Motivation

`no_init_weights` is used in `from_pretrained` to speed up loading large models.
However, torch-built-in modules like `torch.nn.Linear` are heavily used in models of `transformers`, while its weights initialization cannot be disabled by `no_init_weights`.
And in the doc string of `no_init_weights`, it should "globally disable weight initialization".

### Your contribution

possible implementation
```python
class SupportsResetParameters(Protocol):
    def reset_parameters(self): ...

@contextmanager
def no_init(module_classes: Iterable[Type[SupportsResetParameters]]):
    saved = {m: vars(m).get('reset_parameters') for m in module_classes}
    def no_op(_): pass
    for m in saved: m.reset_parameters = no_op # Iterable can only be safely iterated through once
    try:
        yield
    finally:
        for m, init in saved.items():
            del m.reset_parameters
            if init is not None:
                m.reset_parameters = init

TORCH_BUILT_IN_MODULES = [nn.Linear, nn.Conv2d, ...]

@contextmanager
def no_init_weights():
    """
    Context manager to globally disable weight initialization to speed up loading large models.
    """
    global _init_weights
    saved = _init_weights
    _init_weights = False
    try:
        with no_init(TORCH_BUILT_IN_MODULES):
            yield
    finally:
        _init_weights = saved
```

### LINKED ISSUE #26258: Make `_fast_init` fast again (by surely skipping model weights init)!
I observed that loading pre-trained model takes rather long, even when loading cached models from fast SSD. It is especially noticeable when dealing with LLMs with billions of weights.
Apparently, majority of the time is lost [in this section of the code](https://github.com/huggingface/transformers/blob/eb8489971ac1415f67b0abdd1584fde8b659ced9/src/transformers/modeling_utils.py#L2996):
```
# Instantiate model.
    init_contexts = [no_init_weights(_enable=_fast_init)]
# (...)
with ContextManagers(init_contexts):
	model = cls(config, *model_args, **model_kwargs)
```
Time is spent on weights initialization (by `torch.nn.init.kaiming_uniform_()` and similar) is wasted, because the newly initialized weights will be then replaced by loaded ones. The `no_init_weights` context manager sets `_init_weights` global variable, but it gets ignored by model's code (tested on Llama_2_7B).

I recently discussed a similar issue with PEFT team, but there it was easier to solve, because in PEFT the init code was dealing with specific torch.nn layer. see https://github.com/huggingface/peft/issues/871 and linked PRs by @BenjaminBossan. Here we need a model-scale solution.

One (not perfectly elegant one) - is to temporarily override methods like torch.nn.init.kaiming_uniform_().  It is used in [our SpQR repo](https://github.com/Vahe1994/SpQR/blob/39753b62378f0e036de181ea07b20951e0cd2359/modelutils.py#L15): 
```
@contextmanager
def suspend_nn_inits():
    skip = lambda *args, **kwargs: None
    saved_inits = torch.nn.init.kaiming_uniform_, torch.nn.init.uniform_, torch.nn.init.normal_  # saving
    torch.nn.init.kaiming_uniform_ = torch.nn.init.uniform_ = torch.nn.init.normal_ = skip  # replacing
    try:
        yield
    finally:
        torch.nn.init.kaiming_uniform_, torch.nn.init.uniform_, torch.nn.init.normal_ = saved_inits  # restoring
```
but there may be better ones, using some native torch tools? 

I'd be glad to contribute a PR with the maintainers blessing. Summoning @younesbelkada 

### System Info
A100-80G + SSD + mucho RAM and kernels.

### Who can help?
@younesbelkada 

### Reproduction
load model, measure timing for [this line](https://github.com/huggingface/transformers/blob/eb8489971ac1415f67b0abdd1584fde8b659ced9/src/transformers/modeling_utils.py#L2996)

### Expected behavior
faster loading

### PR DESCRIPTION
**Pull request by ArthurZucker**
# what does this PR do 
Skips all layer initialization when loading from pretrained without accelerate. 

From ~20seconds to 5 seconds for a 7B model like llama.

The Weights are effectively initialized in « init_weights_ » of the pretrained method. All internal calls are skipped

- [x] Check that if a linear layer is missing it will be initialized! (Loading `AutoModelForCausalLM` from `AutoModel`

fixes #26258 and fixes #18505

` model = XXXX.from_pretrained(model_id, torch_dtype=torch.float16, low_cpu_mem_usage=True")`  might fail

### CODE REVIEW

**Comment by ArthurZucker** [comment]
- explicit overwrite breaks fx and is not that faster 
- Non initialized is not always zeros (failing tests) just make sure it’s not initialized

File: tests/test_modeling_common.py [hunk 1]
```diff
@@ -427,6 +430,58 @@ class CopyClass(model_class):
                         max_diff = (model_slow_init.state_dict()[key] - model_fast_init.state_dict()[key]).sum().item()
                     self.assertLessEqual(max_diff, 1e-3, msg=f"{key} not identical")
 
+    def test_fast_init_context_manager(self):
+        # 1. Create a dummy class. Should have buffers as well? To make sure we test __init__
+        class MyClass(PreTrainedModel):
+            config_class = PretrainedConfig
+
+            def __init__(self, config=PretrainedConfig()):
```
Comment (LysandreJik) [inline tests/test_modeling_common.py:438]: Wouldn't you rather pass the raw class and instantiate it within the method? Here it'll be instantiated once; if you re-use it you'll get the potential changes done to the class in previous initializations.

Comment (LysandreJik) [inline tests/test_modeling_common.py:438]:
```suggestion
            def __init__(self, config=None):
```

File: tests/test_modeling_common.py [hunk 2]
```diff
@@ -427,6 +430,58 @@ class CopyClass(model_class):
...
+                super().__init__(config)
```
Comment (LysandreJik) [inline tests/test_modeling_common.py:439]:
```suggestion
                super().__init__(config if config is not None else PretrainedConfig())
```

File: src/transformers/modeling_utils.py [hunk 3]
```diff
@@ -164,12 +180,54 @@ def no_init_weights(_enable=True):
...
+            torch.nn.init.uniform = TORCH_INIT_FUNCTIONS["uniform"]
+            torch.nn.init.normal = TORCH_INIT_FUNCTIONS["normal"]
+            torch.nn.init.xavier_uniform = TORCH_INIT_FUNCTIONS["xavier_uniform"]
+            torch.nn.init.xavier_normal = TORCH_INIT_FUNCTIONS["xavier_normal"]
+            torch.nn.init.kaiming_uniform = TORCH_INIT_FUNCTIONS["kaiming_uniform"]
+            torch.nn.init.kaiming_normal = TORCH_INIT_FUNCTIONS["kaiming_normal"]
+        
+            # # Restore the original initialization functions
+            # for name, init_func in TORCH_INIT_FUNCTIONS.items():
+            #     setattr(torch.nn.init, name, init_func)
```
Comment (LysandreJik) [inline src/transformers/modeling_utils.py:230]: Why wasn't this preferred?

**Review by LysandreJik — APPROVED** [review]
Smart, albeit a bit manual :smile: 

LGTM

File: src/transformers/modeling_utils.py (hunk 3 above)
Comment (ArthurZucker) [reply src/transformers/modeling_utils.py:230]: This will be prefered !

File: tests/test_modeling_common.py (hunk 1 above)
Comment (ArthurZucker) [reply tests/test_modeling_common.py]: Yes

**Pull request merged**